physical/zookeeper.go

```diff
@@ -0,0 +1,213 @@
+package physical
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+// ZookeeperBackend is a physical backend that stores data at specific
+// prefix within Zookeeper. It is used in production situations as
+// it allows Vault to run on multiple machines in a highly-available manner.
+type ZookeeperBackend struct {
+	path   string
+	client *zk.Conn
+}
+
+// newZookeeperBackend constructs a Zookeeper backend using the given API client
+// and the prefix in the KV store.
+func newZookeeperBackend(conf map[string]string) (Backend, error) {
+	// Get the path in Zookeeper
+	path, ok := conf["path"]
+	if !ok {
+		path = "vault/"
+	}
+
+	// Ensure path is suffixed and prefixed (zk requires prefix /)
+	if !strings.HasSuffix(path, "/") {
+		path += "/"
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	// Configure the client
+	machines := strings.Split(conf["address"], ",")
+
+	client, _, err := zk.Connect(machines, time.Second)
+	if err != nil {
+		return nil, fmt.Errorf("client setup failed: %v", err)
+	}
+
+	// Setup the backend
+	c := &ZookeeperBackend{
+		path:   path,
+		client: client,
+	}
+	return c, nil
+}
+
+// zookeeper requires nodes to be there before set and get
+func (c *ZookeeperBackend) ensurePath(path string) {
```

Comment:
I'm wondering if there is a way we can opportunistically avoid this. E.g. can we wait to get an error about a missing parent, and then trigger this, instead of doing it on every request? Seems like we could save a lot of round trips